### PR TITLE
added sd-sdr and scale-aware sdr

### PR DESCRIPTION
**Pull request by popcornell**
Will add PESQ loss when i have tested it more extensively.

### CODE REVIEW

**Comment by mpariente** [comment]
Thanks a lot for the PR ! 

Everything related to losses should be integrated into the [`pit_losses_rebased` branch](https://github.com/mpariente/AsSteroid/tree/pit_loss_rebased/asteroid/losses), which provices better API, more use cases and is unit-tested. 

Regardless, there are few points I'd like to discuss: 
* If my understanding of the paper is correct, both scale-dependent SDR (SD-SDR) and scale-invariant SDR (SI-SDR) are scale aware SDRs. 

* I'd like to keep SD-SDR separate (e.g not taken the min over SD-SDR and SNR as suggested could be done in the paper). Plus taking the minimum over pairwise losses sounds weird to me, it would mean the loss functions could different for two source/target pairs in the same example. Taking the minimum of two losses after PIT would make sense though and would easily be implemented in `System`.

I review the PR but the best to do will be to close it and reopen another one on the feature branch.

File: asteroid/engine/losses.py [hunk 1]
```diff
@@ -97,6 +97,95 @@ def get_loss_func_args(self, infos):
...
+    s_target = torch.unsqueeze(source, dim=1)
+    s_estimate = torch.unsqueeze(est_source, dim=2)
+
+    # compute alpha
+    scale_factor =  torch.sum(s_estimate * s_target, dim=3, keepdim=True)
+    s_target_energy = torch.sum(s_target ** 2, dim=3, keepdim=True) + EPS
+    alpha = (scale_factor / (s_target_energy + EPS))
+
+    # compute snr
+    s_target = s_target.repeat(1, s_target.shape[2], 1, 1)
```
[the remark this answers is not in the document]
Comment (popcornell) [reply asteroid/engine/losses.py]: Ok i kept it because also in the original pairwise_sisdr for scale_invariant=False it was doing the same.

File: asteroid/engine/losses.py [hunk 2]
```diff
@@ -97,6 +97,95 @@ def get_loss_func_args(self, infos):
...
+    pair_wise_proj = scale_factor * s_target / s_target_energy
+
+    # compute snr
+    e_noise = s_estimate - s_target.repeat(1, s_target.shape[2], 1, 1)
+
+    # [batch, n_src, n_src]
+    pair_wise_snr = torch.sum(pair_wise_proj ** 2, dim=3) / (
+            torch.sum(e_noise ** 2, dim=3) + EPS)
+
+    return -10 * torch.log10(pair_wise_snr + EPS)
```
[the remark this answers is not in the document]
Comment (popcornell) [reply asteroid/engine/losses.py]: Yes it should be

**Comment by popcornell** [comment]
> If my understanding of the paper is correct, both scale-dependent SDR (SD-SDR) and scale-invariant SDR (SI-SDR) are scale aware SDRs.

Well the SI-SDR should be insensitive to rescaling right ? So it does not account for both up-scaling and down-scaling errors. While SD-SDR accounts for downscaling and partly for up-scaling. 
But again it is true that unlike plain SNR it prevents the network to use scaling tricks to artificially boost the SNR. I called it that way because it was the first thing it came into my mind this afternoon. 

> I'd like to keep SD-SDR separate (e.g not taken the min over SD-SDR and SNR as suggested could be done in the paper). Plus taking the minimum over pairwise losses sounds weird to me, it would mean the loss functions could different for two source/target pairs in the same example. Taking the minimum of two losses after PIT would make sense though and would easily be implemented in System.

Yes you are right probably it is better to move it afterwards. 

Ok i will then remove this PR and apply the changes then re-open it on the losses branch.

**Comment by mpariente** [comment]
Hey @popcornell , if you want to make a PR against master for the loss, that'd be great :-)

**Comment by popcornell** [comment]
Yep, now that the interface for the losses is ready I will PR to the the master. 
Guess now i have to take care for both nosrc and pairwise right ?

**Comment by mpariente** [comment]
Of course, if it covers all the modes, it's great ! Plus, it enables us to unit test that everything is fine.
Otherwise, you can just do the ones you feel like, I can probably make the rest, as you want. 
Thanks !
